### ISSUE TITLE
[feature request] Programmatic usage support

### ISSUE THREAD

**Issue by lqmanh** [sinclairzx81/typebox-workbench#1]
Hi,

I just recently came across this project and I really love it. Still, can I somehow utilize this TypeScript -> Typebox feature via programmatic usage? (similar to https://github.com/fabien0102/ts-to-zod)

Thank you for your great work.

**Comment by qurafi** [comment]
+1

I found out the ts to json schema is more mature and fast compared to tools like:
- https://github.com/YousefED/typescript-json-schema
- https://github.com/vega/ts-json-schema-generator

**Comment by sinclairzx81** [comment]
@lqmanh @mhmd-22 Hi! thanks :)

The workbench is currently using a WIP codegen backend package written to handle TS transformations. This package is yet to be published (and is currently manually integrated into the workbench for testing), but there's actually another project handling more traditional CLI / Programmatic use cases. I would suggest this if you're looking for integrating more traditional codegen into your workflows.

You can find the project here https://github.com/xddq/ts2typebox with programmatic usage implemented here https://github.com/xddq/ts2typebox/blob/main/src/programmatic-usage.ts. As mentioned, this package is using the same compiler backend as the workbench, so CLI/programmatic output should match what you see in the workbench.

Hope this helps!
S

CC @xddq

**Comment by qurafi** [comment]
That's great news. I'm currently working on a vite/rollup plugin that compile json schema files to validation code via Ajv to use with Svelte. And I'm using json-schema-to-typescript to generate types but In my opinion I think the opposite way is much more enjoyable to write👍

**Comment by xddq** [comment]
@lqmanh Basically what @sinclairzx81 already said.

@mhmd-22 mentioned an issue with the ergonomics of the exports/imports. I will fix this and also adapt the pure programmatic usage API to be similar to what I described [here](https://github.com/xddq/ts2typebox/issues/19#issuecomment-1565966782). Examples and tests for the programmatic usage will then also be added in the same go.

Open to feedback on the initial idea of the API change. : ]

**Comment by xddq** [comment]
The new api changes for https://github.com/xddq/ts2typebox are now available with the latest version (1.4.0).

@lqmanh (If you got everything you need) don't forget to close the issue

**Comment by lqmanh** [comment]
@xddq Thanks, I'll take a look

**Comment by sinclairzx81** [comment]
Going to close this one off.

Cheers all

**Comment by qurafi** [comment]
@sinclairzx81 @xddq  @lqmanh 
For anyone interested in the project I'm working on. I made it public at https://github.com/qurafi/ajv-tools. It's a rollup/vite plugin to manage Ajv compilations. 

As for ts2typebox feature, I had to abandon it because the Typescript's Typechecker API is noticeably slow, it takes 200ms+ on a single file. Maybe I'm doing something wrong but I might give it more time later.
